### PR TITLE
primTorch aliases

### PR DESCRIPTION
**Pull request by mruberry**
primTorch aliases allow aliasing relationships between PyTorch operators to be expressed programmatically. The appropriate bindings and C++ for the alias operations are generated at build time, and the relationships are programmatically inspectable at runtime. This advances the primTorch's projects goal of clarifying the relationships between PyTorch operators to systems that consume them while preserving a single source of operator truth and making them easier to implement and maintain. 

In primTorch alias relationships are expressed in tools/codegen/prim/aliases.py. At build time this file is copied to torch/prim/aliases.py and can be imported from the `torch.prim` module. A typical alias relationship is expressed as:

```
AliasInfo('absolute', alias_for='abs'),
```

although there are options for creating aliases in namespaces or creating aliases with or without method variants. At build time these relationships are reviewed and native function data structures augmented and C++ generated to implement the alias relationship, at which point the build system proceeds as usual. The generated C++ appears in aten/src/ATen/native/Aliases.cpp.

Although primTorch generates the bindings and C++ for aliases it does not do everything. Users still have to update the docs and overrides.py like for every operator, and they need to update the  alias map in torch/csrc/jit/passes/normalize_ops.cpp manually. They also have to ensure the alias relationship is described by an OpInfo, and there is a test to verify it is. 

In the near future more (all?) aliases will be expressed as primTorch aliases and additional operators will also be programmatically inspectable at runtime. It would also be nice if the overrides.py updates weren't required for operators with OpInfos, but that's a separate issue. 

cc @zou3519 @kshitij12345

### CODE REVIEW

File: tools/codegen/prim/alias_gen.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        call_args.append(args.self_arg.argument)
+    call_args.extend(args.post_self_positional)
+    call_args.extend(args.pre_tensor_options_kwarg_only)
+    call_args.extend(args.post_tensor_options_kwarg_only)
+
+    arg_cpp = ", ".join(arg.name for arg in call_args)
+    return "at::" + fn_cpp + "(" + arg_cpp + ")"
+
+# Generates the C++ for a function signature
+def generate_signature_cpp(return_cpp: str, fn_cpp: str, args: Arguments) -> str:
```
Comment (ngimel) [inline tools/codegen/prim/alias_gen.py:115]: `gen.py` already has machinery to generate cpp signature from native function object, see e.g. `compute_registration_declarations` that does that (+ comments that we don't need here), it's better to reuse it 
https://github.com/pytorch/pytorch/blob/71a41323bb42e3b75d259a546a31172d5fe6590c/tools/codegen/gen.py#L914

File: aten/src/ATen/native/Aliases.cpp.in [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#include <ATen/ATen.h>
+
+namespace at { namespace native {
+
+$alias_definitions
```
Comment (zou3519) [inline aten/src/ATen/native/Aliases.cpp.in:5]: Could we see some sample generated code?

**Review by zou3519 — COMMENTED** [review]
How does this approach compare to https://github.com/pytorch/pytorch/pull/63349 , where aliases are just C++ functions that call the original function?

**Comment by mruberry** [comment]
> How does this approach compare to #63349 , where aliases are just C++ functions that call the original function?

That's why the aliases are here, too.

File: aten/src/ATen/native/Aliases.cpp.in (hunk 2 above)
Comment (mruberry) [reply aten/src/ATen/native/Aliases.cpp.in:5]: The template parameter is replaced with:

```
Tensor absolute(const Tensor& self){
	return at::abs(self);
}
Tensor& absolute_(Tensor& self){
	return at::abs_(self);
}
Tensor& absolute_out(const Tensor& self, Tensor& out){
	return at::abs_out(out, self);
}
Tensor arccos(const Tensor& self){
	return at::acos(self);
}
Tensor& arccos_(Tensor& self){
	return at::acos_(self);
}
Tensor& arccos_out(const Tensor& self, Tensor& out){
	return at::acos_out(out, self);
}
Tensor concat(TensorList tensors, int64_t dim){
	return at::cat(tensors, dim);
}
Tensor& concat_out(TensorList tensors, int64_t dim, Tensor& out){
	return at::cat_out(out, tensors, dim);
}
Tensor concat_names(TensorList tensors, Dimname dim){
	return at::cat(tensors, dim);
}
Tensor& concat_names_out(TensorList tensors, Dimname dim, Tensor& out){
	return at::cat_out(out, tensors, dim);
}
Tensor clip(const Tensor& self, const c10::optional<Scalar>& min, const c10::optional<Scalar>& max){
	return at::clamp(self, min, max);
}
Tensor clip_Tensor(const Tensor& self, const c10::optional<Tensor>& min, const c10::optional<Tensor>& max){
	return at::clamp(self, min, max);
}
Tensor& clip_(Tensor& self, const c10::optional<Scalar>& min, const c10::optional<Scalar>& max){
	return at::clamp_(self, min, max);
}
Tensor& clip__Tensor(Tensor& self, const c10::optional<Tensor>& min, const c10::optional<Tensor>& max){
	return at::clamp_(self, min, max);
}
Tensor& clip_out(const Tensor& self, const c10::optional<Scalar>& min, const c10::optional<Scalar>& max, Tensor& out){
	return at::clamp_out(out, self, min, max);
}
Tensor& clip_Tensor_out(const Tensor& self, const c10::optional<Tensor>& min, const c10::optional<Tensor>& max, Tensor& out){
	return at::clamp_out(out, self, min, max);
}
Tensor moveaxis_intlist(const Tensor& self, IntArrayRef source, IntArrayRef destination){
	return at::movedim(self, source, destination);
}
Tensor moveaxis_int(const Tensor& self, int64_t source, int64_t destination){
	return at::movedim(self, source, destination);
}
Tensor& less_Scalar_out(const Tensor& self, const Scalar& other, Tensor& out){
	return at::lt_out(out, self, other);
}
Tensor less_Scalar(const Tensor& self, const Scalar& other){
	return at::lt(self, other);
}
Tensor& less_Tensor_out(const Tensor& self, const Tensor& other, Tensor& out){
	return at::lt_out(out, self, other);
}
Tensor less_Tensor(const Tensor& self, const Tensor& other){
	return at::lt(self, other);
}
Tensor& less__Scalar(Tensor& self, const Scalar& other){
	return self.lt_(other);
}
Tensor& less__Tensor(Tensor& self, const Tensor& other){
	return self.lt_(other);
}
Tensor det(const Tensor& self){
	return at::linalg_det(self);
}
Tensor& det_out(const Tensor& self, Tensor& out){
	return at::linalg_det_out(out, self);
}
```

Comment (zou3519) [reply aten/src/ATen/native/Aliases.cpp.in]: Thanks for the code, this is cool!

**Pull request closed without merging**